src/cursor/cur_join.c

```diff
@@ -1104,19 +1103,21 @@ __curjoin_next(WT_CURSOR *cursor)
 
 	if (ret == 0) {
 		/*
-		 * Position the 'main' cursor, this will be used to
-		 * retrieve values from the cursor join.
+		 * Position the 'main' cursor, this will be used to retrieve
+		 * values from the cursor join.  The key we have is raw, but
+		 * the values we retrieve may not be raw.  So we need to
+		 * turn on the raw interface temporarily.
 		 */
 		c = cjoin->main;
-		if (WT_CURSOR_RECNO(cursor) &&
-		    !F_ISSET(cursor, WT_CURSTD_RAW)) {
-			p = (const uint8_t *)iter->curkey->data;
-			WT_ERR(__wt_vunpack_uint(&p, iter->curkey->size,
-			    &cjoin->iface.recno));
-			c->set_key(c, cjoin->iface.recno);
-		} else
-			c->set_key(c, iter->curkey);
-		WT_ERR(c->search(c));
+		F_SET(c, WT_CURSTD_RAW);
+		c->set_key(c, iter->curkey);
+		if (!F_ISSET(cursor, WT_CURSTD_RAW))
+			F_CLR(c, WT_CURSTD_RAW);
```

Comment:
@ddanderson, we have `__wt_cursor_set_raw_key` that should take care of these 4 lines for you:
```c
    __wt_cursor_set_raw_key(c, iter->curkey);
```